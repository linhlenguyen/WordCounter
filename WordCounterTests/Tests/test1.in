1 What are functional programming languages?
Programming languages such as C/C++/Java/Python are called imperative programming languages because they consist of sequences of actions. The programmer quite explicitly tells the computer how to perform a task, step-by-step. Functional programming languages work differently. Rather than performing actions in a sequence, they evaluate expressions.

1.1 The level of abstraction
There are two areas that are fundamental to programming a computer - resource management and sequencing. Resource management (allocating registers and memory) has been the target of vast abstraction, most new languages (imperative as well as functional) have implemented garbage collection to remove resource management from the problem, and lets the programmer focus on the algorithm instead of the book-keeping task of allocating memory. Sequencing has also undergone some abstraction, although not nearly to the same extent. Imperative languages have done so by introducing new keywords and standard libraries. For example, most imperative languages have special syntax for constructing several slightly different loops, you no longer have to do all the tasks of managing these loops yourself. But imperative languages are based upon the notion of sequencing - they can never escape it completely. The only way to raise the level of abstraction in the sequencing area for an imperative language is to introduce more keywords or standard functions, thus cluttering up the language. This close relationship between imperative languages and the task of sequencing commands for the processor to execute means that imperative languages can never rise above the task of sequencing, and as such can never reach the same level of abstraction that functional programming languages can.

In Haskell, the sequencing task is removed. You only care what the program is to compute not how or when it is computed. This makes Haskell a more flexible and easy to use language. Haskell tends to be part of the solution for a problem, not a part of the problem itself.
a a a a a a
