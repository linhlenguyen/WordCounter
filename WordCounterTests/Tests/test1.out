the, 20
of, 14
a, 12
to, 11
languages, 11
imperative, 7
and, 7
sequencing, 6
task, 5
abstraction, 5
programming, 5
are, 5
as, 4
can, 4
for, 4
have, 4
is, 4
Haskell, 3
that, 3
problem, 3
management, 3
language, 3
functional, 3
The, 3
not, 3
never, 3
they, 3
level, 3
most, 2
it, 2
only, 2
or, 2
has, 2
memory, 2
allocating, 2
same, 2
keywords, 2
standard, 2
resource, 2
more, 2
new, 2
This, 2
loops, 2
programmer, 2
computer, 2
how, 2
step, 2
actions, 2
C, 2
such, 2
by, 2
in, 2
part, 2
managing, 1
way, 1
makes, 1
completely, 1
flexible, 1
escape, 1
easy, 1
use, 1
no, 1
tends, 1
notion, 1
be, 1
longer, 1
upon, 1
do, 1
based, 1
But, 1
yourself, 1
these, 1
all, 1
tasks, 1
raise, 1
above, 1
an, 1
reach, 1
means, 1
execute, 1
processor, 1
commands, 1
In, 1
you, 1
removed, 1
You, 1
between, 1
care, 1
area, 1
relationship, 1
close, 1
up, 1
program, 1
cluttering, 1
thus, 1
functions, 1
compute, 1
when, 1
introduce, 1
computed, 1
rise, 1
what, 1
What, 1
although, 1
slightly, 1
target, 1
been, 1
registers, 1
Resource, 1
fundamental, 1
areas, 1
two, 1
There, 1
expressions, 1
evaluate, 1
sequence, 1
performing, 1
than, 1
vast, 1
Rather, 1
work, 1
Functional, 1
perform, 1
tells, 1
explicitly, 1
quite, 1
sequences, 1
consist, 1
because, 1
called, 1
Python, 1
Java, 1
Programming, 1
differently, 1
different, 1
well, 1
garbage, 1
several, 1
constructing, 1
syntax, 1
special, 1
example, 1
For, 1
libraries, 1
introducing, 1
so, 1
done, 1
Imperative, 1
extent, 1
nearly, 1
implemented, 1
solution, 1
undergone, 1
also, 1
Sequencing, 1
keeping, 1
book, 1
instead, 1
algorithm, 1
on, 1
focus, 1
lets, 1
from, 1
remove, 1
collection, 1
some, 1
itself, 1

